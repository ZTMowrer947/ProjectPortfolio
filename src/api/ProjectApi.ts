// Imports
import { projects as projectData } from '@/../data.json';
import Project from '@/models/Project';

// Project data
const projects: Project[] = projectData.map((project) => ({
  id: project.id,
  name: project.project_name,
  description: project.description,
  technologies: project.technologies,
  liveLink: project.live_link ?? null,
  githubLink: project.github_link,
  landingImageUrl: project.image_urls.landing,
  galleryImageUrls: project.image_urls.gallery,
}));

// Service
class ProjectApi {
  public static async getList(): Promise<Project[]> {
    return Promise.resolve(projects);
  }

  public static async get(id: number): Promise<Project | undefined> {
    return Promise.resolve(projects.find((project) => project.id === id));
  }
}

// Exports
export default ProjectApi;
