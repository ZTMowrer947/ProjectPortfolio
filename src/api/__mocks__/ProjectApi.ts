// Imports
import faker from 'faker';

import Project from '@/models/Project';

// Mock data
const projects = Array.from(
  { length: 10 },
  (_, index): Project => ({
    id: index + 1,
    name: faker.random.words(3),
    description: faker.lorem.paragraph(),
    technologies: Array.from({ length: 3 }, () => faker.random.word()),
    liveLink: faker.internet.url(),
    githubLink: faker.internet.url(),
    landingImageUrl: faker.image.imageUrl(),
    galleryImageUrls: Array.from({ length: 3 }, () => faker.image.imageUrl()),
  })
);

// Mock Service
class ProjectMockApi {
  public static async getList(): Promise<Project[]> {
    return Promise.resolve(projects);
  }

  public static async get(id: number): Promise<Project | undefined> {
    return Promise.resolve(projects.find((project) => project.id === id));
  }
}

// Exports
export default ProjectMockApi;
