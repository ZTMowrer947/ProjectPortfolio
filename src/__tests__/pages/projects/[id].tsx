// Imports
import { render } from '@testing-library/react';

import ProjectApi from '@/api/ProjectApi';
import ProjectDetail from '@/pages/projects/[id]';

// Jest mocks
jest.mock('@/api/ProjectApi');

// Test Suite
describe('<ProjectDetail>', () => {
  it('should render project details', async () => {
    // Retrieve mock project data
    const project = await ProjectApi.get(1).toPromise();

    if (!project) throw new Error('Could not retrieve project');

    // Render component
    const { queryByText, queryByTestId, queryAllByTestId } = render(
      <ProjectDetail project={project} />
    );

    // Retrieve project detail elements
    const projectHeading = queryByText(project.name);
    const projectDescription = queryByText(project.description);
    const projectTechnologies = queryAllByTestId('project-technology');
    const projectLiveLink = queryByTestId('live-link');
    const projectSourceLink = queryByTestId('gh-link');
    const projectGalleryImages = queryAllByTestId('gallery-image');

    // Expect all single elements to exist in the docment
    expect(projectHeading).toBeInTheDocument();
    expect(projectDescription).toBeInTheDocument();
    expect(projectLiveLink).toBeInTheDocument();
    expect(projectSourceLink).toBeInTheDocument();

    // Expect technology and gallery images to match length of corresponding project items
    expect(projectTechnologies).toHaveLength(project.technologiesUsed.length);
    expect(projectGalleryImages).toHaveLength(project.galleryImages.length);

    // For each technology item,
    projectTechnologies.forEach((techItem, index) => {
      // Get corresponding technology
      const technologies = project.technologiesUsed[index];

      // Expect each technology item to contain name of technology
      expect(techItem).toHaveTextContent(technologies);
    });

    // For each gallery image,
    projectGalleryImages.forEach((galleryImage, index) => {
      // Get corresponding expected URL
      const galleryImageUrl = project.galleryImages[index];

      // Expect gallery image to have correct source URL
      expect(galleryImage).toHaveAttribute('src', galleryImageUrl);
    });
  });
});
