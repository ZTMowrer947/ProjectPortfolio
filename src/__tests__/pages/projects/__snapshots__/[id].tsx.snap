// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetail page should render all the details of a given project, including its live link if it has one 1`] = `
Array [
  <header
    className="mt-3"
  >
    <a
      className="text-decoration-none"
      href="/projects"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      ← Back
    </a>
  </header>,
  <hr />,
  <article
    className="portfolio-intro portfolio-project"
  >
    <div
      className="row"
    >
      <div
        className="col-md-9 col-12"
      >
        <h1>
          Project #1
        </h1>
        <div
          className="lead text-dark"
        >
          <p>
            Lorem ipsum dolor sit amet.
          </p>
        </div>
      </div>
      <div
        className="col-md-3 col-12"
      >
        <h6
          className="text-uppercase"
        >
          Technologies
        </h6>
        <ul>
          <li
            data-testid="project-technology"
          >
            Jest
          </li>
        </ul>
        <a
          aria-label="live-link"
          className="btn btn-dark btn-block text-nowrap"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Live Demo
        </a>
        <a
          aria-label="source-link"
          className="btn btn-dark btn-block text-nowrap"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Github Repo
        </a>
      </div>
    </div>
    <hr />
    <img
      alt="Gallery image for \\"Project #1\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
    <img
      alt="Gallery image for \\"Project #1\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
    <img
      alt="Gallery image for \\"Project #1\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
  </article>,
]
`;

exports[`ProjectDetail page should render all the details of a given project, save for its live link if it doesn't have one 1`] = `
Array [
  <header
    className="mt-3"
  >
    <a
      className="text-decoration-none"
      href="/projects"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      ← Back
    </a>
  </header>,
  <hr />,
  <article
    className="portfolio-intro portfolio-project"
  >
    <div
      className="row"
    >
      <div
        className="col-md-9 col-12"
      >
        <h1>
          Project #2
        </h1>
        <div
          className="lead text-dark"
        >
          <p>
            Lorem ipsum dolor sit amet.
          </p>
        </div>
      </div>
      <div
        className="col-md-3 col-12"
      >
        <h6
          className="text-uppercase"
        >
          Technologies
        </h6>
        <ul>
          <li
            data-testid="project-technology"
          >
            Jest
          </li>
        </ul>
        <a
          aria-label="source-link"
          className="btn btn-dark btn-block text-nowrap"
          href="http://example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          Github Repo
        </a>
      </div>
    </div>
    <hr />
    <img
      alt="Gallery image for \\"Project #2\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
    <img
      alt="Gallery image for \\"Project #2\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
    <img
      alt="Gallery image for \\"Project #2\\""
      className="img-fluid img-thumbnail"
      data-testid="gallery-image"
      src="https://place-hold.it/640x480"
    />
  </article>,
]
`;
